nautobot_ssot/jobs/base.py

```diff
@@ -33,31 +36,135 @@
+
+        - self.kwargs     (corresponds to the Job's `data` input, including 'dry_run' option)
+        - self.job_result (as per Job API)
+        """
+        raise NotImplementedError
+
+    def calculate_diff(self):
+        """Method to calculate the difference from SOURCE to TARGET adapter and store in `self.diff`.
+
+        This is a generic implementation that you could overwrite completely in you custom logic.
+        """
+        if self.source_adapter and self.target_adapter:
+            self.diffsync_flags = (
+                DiffSyncFlags.CONTINUE_ON_FAILURE
+                | DiffSyncFlags.LOG_UNCHANGED_RECORDS
+                | DiffSyncFlags.SKIP_UNMATCHED_DST
+            )
+            self.diff = self.source_adapter.diff_to(self.target_adapter, flags=self.diffsync_flags)
+            self.sync.diff = self.diff.dict()
+            self.sync.save()
+
+    def execute_sync(self):
+        """Method to synchronize the difference from `self.diff`, from SOURCE to TARGET adapter.
+
+        This is a generic implementation that you could overwrite completely in you custom logic.
+        """
+        if self.source_adapter and self.target_adapter:
+            self.source_adapter.sync_to(self.target_adapter, flags=self.diffsync_flags)
 
     def sync_data(self):
-        """Method to be implemented by data sync concrete Job implementations.
+        """Method to load data from adapters, calculate diffs and sync (if not dry-run).
+
+        It is composed by 4 methods:
+        - self.load_source_adapter: instantiates the source adapter (self.source_adapter) and loads its data
+        - self.load_target_adapter: instantiates the target adapter (self.target_adapter) and loads its data
+        - self.calculate_diff: generates the diff from source to target adapter and stores it in self.diff
+        - self.execute_sync: if not dry-run, uses the self.diff to synchronize from source to target
 
+        This is a generic implementation that you could overwrite completely in you custom logic.
         Available instance attributes include:
 
         - self.kwargs     (corresponds to the Job's `data` input, including 'dry_run' option)
         - self.commit     (should generally be True)
         - self.sync       (Sync instance tracking this job execution)
         - self.job_result (as per Job API)
         """
-        pass  # pylint: disable=unnecessary-pass
+        if not self.sync:
+            return
+
+        if self.kwargs["memory_profiling"]:
+            tracemalloc.start()
+
+        start_time = datetime.now()
+
+        self.log_info(message="Loading current data from source adapter...")
+        self.load_source_adapter()
+        load_source_adapter_time = datetime.now()
+        self.sync.load_time_source = load_source_adapter_time - start_time
+        self.sync.save()
+        self.log_info(
+            message=f"Source Load Time from {self.source_adapter.__class__.__name__}: {self.sync.load_time_source}"
```

Comment:
DiffSync adapters have a perfectly good `__str__` implementation:

```suggestion
            message=f"Source Load Time from {self.source_adapter}: {self.sync.load_time_source}"
```